src/lib/car/index.ts

```diff
@@ -0,0 +1,201 @@
+ */
+export interface IndexEntry {
+  multihash: Uint8Array;
+  offset: number;
+}
+
+/**
+ * Writes a CARv1 file from a map of CIDs to block data, and returns the raw bytes and index entries.
+ *
+ * @param nodes - A map of CID string to corresponding block data.
+ * @param rootCID - The root CID of the DAG to be used in the CAR file header.
+ * @returns A Promise resolving to an object containing the CAR file bytes and an array of multihash index entries.
+ */
+export async function writeCarFileWithOffsets(
+  nodes: Map<string, Uint8Array>,
+  rootCID: CID,
+): Promise<{ carBytes: Uint8Array; indexEntries: IndexEntry[] }> {
+  const { writer, out } = await CarWriter.create([rootCID]);
+  let offset = CarBufferWriter.headerLength({ roots: [rootCID] });
+
+  const blockQueue = createBlockQueue(nodes);
+  const indexEntries = new Array<IndexEntry>(blockQueue.length);
+  const writtenBlocks = new Set<string>();
+  const carChunks: Uint8Array[] = [];
+
+  const collectPromise = collectChunksAndOffsets(
+    out,
+    blockQueue,
+    writtenBlocks,
+    indexEntries,
+    carChunks,
+    () => offset,
+    (len) => {
+      offset += len;
+    },
+  );
+
+  for (const [cidStr, bytes] of nodes.entries()) {
+    const cid = CID.parse(cidStr);
+    await writer.put({ cid, bytes });
+  }
+
+  await writer.close();
+  await collectPromise;
+
+  const carBytes = concatChunks(carChunks);
+
+  return {
+    carBytes,
+    indexEntries,
+  };
+}
+
+/**
+ * Constructs a block queue from the input node map, preserving insertion order.
+ *
+ * @param nodes - A map of CID string to block data.
+ * @returns An array of [CID string, CID] pairs.
+ */
+function createBlockQueue(nodes: Map<string, Uint8Array>): [string, CID][] {
+  const blockQueue = new Array<[string, CID]>(nodes.size);
+  let i = 0;
+  for (const key of nodes.keys()) {
+    blockQueue[i++] = [key, CID.parse(key)];
+  }
+  return blockQueue;
+}
+
+/**
+ * Collects chunks emitted from the CAR writer output and tracks block offsets to build an index.
+ *
+ * @param out - The async iterable stream of CAR chunks.
+ * @param blockQueue - Array of [CID string, CID] representing the expected block write order.
+ * @param writtenBlocks - A Set tracking which blocks have been indexed already.
+ * @param indexEntries - The array to store offset information for each block.
+ * @param carChunks - Accumulator array for CAR chunks.
+ * @param getOffset - Function returning the current byte offset.
+ * @param advanceOffset - Function to advance the byte offset after each chunk.
+ */
+async function collectChunksAndOffsets(
+  out: AsyncIterable<Uint8Array>,
+  blockQueue: [string, CID][],
+  writtenBlocks: Set<string>,
+  indexEntries: IndexEntry[],
+  carChunks: Uint8Array[],
+  getOffset: () => number,
+  advanceOffset: (len: number) => void,
+): Promise<void> {
+  let blockIndex = 0;
+
+  for await (const chunk of out) {
+    carChunks.push(chunk);
+
+    while (blockIndex < blockQueue.length) {
```

Comment:
Wouldn't a for loop be more readable? You can even use `entries` if you need the index